libnd4j/include/ops/declarable/generic/random/multinomial.cpp

```diff
@@ -0,0 +1,107 @@
+ * SPDX-License-Identifier: Apache-2.0
+ ******************************************************************************/
+
+//
+// @author Oleh Semeniv (oleg.semeniv@gmail.com)
+//
+
+#include <op_boilerplate.h>
+#if NOT_EXCLUDED(OP_random_multinomial)
+
+#include <ops/declarable/CustomOperations.h>
+#include <helpers/RandomLauncher.h>
+#include <ops/declarable/helpers/random.h>
+
+namespace nd4j {
+    namespace ops {
+        ///////////////////////
+        /**
+         * multinomial (categorical) random generator
+         * takes 2D ndarray with logits with shape [batch_size (N), num_classes (K)]
+         * represents the unnormalized log-probabilities for all classes.
+         * Int arguments: 0 - scalar value of samples number, number of independent samples to draw for each experiment 1,N.
+         * Int arguments: 1 - optional argument, corresponds to dimension with batch_size
+         * Int arguments: 2 - optional argument, integer type to use for the output. Default int64.
+         */
+         // used https://en.wikipedia.org/wiki/Categorical_distribution
+         // methods: gumbel trick + softmax + argmax
+        CUSTOM_OP_IMPL(random_multinomial, 1, 1, false, 0, 1) {
+            
+            auto input = INPUT_VARIABLE(0);
+            auto output = OUTPUT_VARIABLE(0);
+            auto nSamples = INT_ARG(0);
+            const int argSize = block.getIArguments()->size();
+
+            REQUIRE_TRUE(argSize >= 1, 0, "RANDOM_MULTINOMIAL OP: Have to be specified at least one sample,"
+                                          " but got no argumets instead.");
+
+            // do nothing if number of samples = 0
+            if (0 == nSamples) 
+                return Status::OK();
+
+            REQUIRE_TRUE(!input->isEmpty(), 0, "RANDOM_MULTINOMIAL OP: Have to be provided at least one logits. ");
+            REQUIRE_TRUE(nSamples > 0, 0, "RANDOM_MULTINOMIAL OP: Number of samples should be greater then 0, got %i. ", nSamples);
+
+            const int rank = input->rankOf();
+            REQUIRE_TRUE(rank == 2, 0, "RANDOM_MULTINOMIAL OP: Logits should be a matrix with rank = 2, but got instead rank = %i.", rank);
+            const int dimC = argSize > 1 ? (INT_ARG(1) >= 0 ? INT_ARG(1) : INT_ARG(1) + rank) : rank - 1;
+
+            auto dimA = (0 == dimC) ? 1 : 0;
+            if (1 == input->sizeAt(dimA)) {
+                *output = 0;
+                return Status::OK();
+            }
+
+            auto rng = block.randomGenerator();
+            helpers::fillRandomMultiNomial(block.launchContext(), rng, *input, *output, dimC);
+            return Status::OK();
+        }
+
+
+        DECLARE_SHAPE_FN(random_multinomial) {
+
+            const int argSize = block.getIArguments()->size();
+            REQUIRE_TRUE(argSize >= 1, 0, "RANDOM_MULTINOMIAL OP: Have to be specified at least one sample,"
+                                          " but got no argumets instead.");
+
+            auto nSamples = INT_ARG(0);
```

Comment:
According to what i see in op spec, nSamples in TF is dynamic, so it should be mapped as another INPUT_VARIABLE

Reply:
https://github.com/KonduitAI/deeplearning4j/blob/b5f0ec072f3fd0da566e32f82c0e43ca36553f39/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/resources/ops.proto#L18430-L18433

Reply:
done